decidim-templates/app/commands/decidim/templates/admin/questionnaire_copier.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+module Decidim
+  module Templates
+    # All the business logic to duplicate a questionnaire
+    module Admin
+      module QuestionnaireCopier
+        def copy_questionnaire_questions(original_questionnaire, new_questionnaire)
+          # start by copying the questions so that they already exist when cross referencing them in the conditions
+          original_questionnaire.questions.each do |original_question|
+            new_question = original_question.dup
+            new_question.questionnaire = new_questionnaire
+            new_question.save!
+            copy_questionnaire_answer_options(original_question, new_question)
+            copy_questionnaire_matrix_rows(original_question, new_question)
+          end
+          original_questionnaire.questions.zip(new_questionnaire.questions).each do |original_question, new_question|
+            copy_question_display_conditions(original_question, new_question)
+          end
+        end
+
+        def copy_questionnaire_answer_options(original_question, new_question)
+          original_question.answer_options.each do |original_answer_option|
+            new_answer_option = original_answer_option.dup
+            new_answer_option.question = new_question
+            new_answer_option.save!
+          end
+        end
+
+        def copy_questionnaire_matrix_rows(original_question, new_question)
+          original_question.matrix_rows.each do |original_matrix_row|
+            new_matrix_row = original_matrix_row.dup
+            new_matrix_row.question = new_question
+            new_matrix_row.save!
+          end
+        end
+
+        def copy_question_display_conditions(original_question, destination_question)
```

Comment:
⚠️ I'm a bit concerned about the complexity of this method. It's executed once per each question and has a loop that runs once per each condition that performs two queries to the database. I think it would be possible to load all questions first and then search in memory instead of doing more requests to the database. Regarding the answers, I'm not so sure what is the best approach, as doing the same could increase too much the complexity of the code. What do you think?

Reply:
@leio10 ok, I've preloaded all the questions to be copied from the original questionnaire (and all associated items). And also implemented in memory searches for the two queries that you were worried about.